lib/Corres_Method.thy

```diff
@@ -0,0 +1,212 @@
+       goals. It is very likely that the method will stop after this and not have solved the goal,
+       but it also very likely that the first thing we want to do for such a goal is clarsimp. That
+       means, overall we might solve a few more goals, and not be detrimental to interactive proof
+       either. *)
+    | fails \<open>is_corres\<close>, fails \<open>is_wp\<close>,
+      clarsimp cong: cong simp del: simp_del simp: simp split del: if_split split_del split: split
+               intro!: intro
+    (* if (and only if) we get to the state where all corres goals and side conditions are solved,
+       attempt to solve all wp goals that were generated in order. We are not using then_all_new_fwd
+       here, because we should only start solving wp goals once *all* corres goals are solved,
+       because otherwise the goal will still have schematic post conditions. Fail if there is a
+       free schematic postcondition despite all these measures.
+       *)
+    | succeeds \<open>is_wp\<close>, fails \<open>is_hoare_schematic_post\<close>,
+      (wpsimp wp: wp wp_del: wp_del simp: simp simp_del: simp_del split: split split_del: split_del
+              cong: cong)+
+   )+)[1]
+
+(* Instance of the corres' method with default cleanup tactic *)
+method corres
+  uses simp term_simp simp_del split split_del cong intro wp wp_del fallback
+  declares corres corres_term corres_splits =
+  corres' \<open>fail\<close> simp: simp term_simp: term_simp simp_del: simp_del split: split
+                 split_del: split_del cong: cong intro: intro wp: wp wp_del: wp_del
+                 fallback: fallback
+                 corres: corres corres_term: corres_term corres_splits: corres_splits
+
+
+section \<open>Corres rule setup\<close>
+
+lemmas [corres_splits] =
+  corres_split
+  corres_splitEE
+
+lemmas corres_split_liftE_bindE [corres_splits] =
+  corres_splitEE[OF corres_liftE_rel_sum[THEN iffD2], simplified]
+
+(* corres_term are rules that are safe when all side conditions can be solved immediately -- they
+   might have guards like \<top> that are too weak in general, but if the goal can be solved with
+   that weak guard, the rule is safe. This enables us to solve trivial cases without adding
+   unsafe rules to the [corres] set. *)
+lemmas [corres_term] =
+  corres_return_eq_same corres_gets_trivial select_corres_eq
+  corres_underlying_assert_assert
+
+lemmas corres_returnOk_eq_same[corres_term] = corres_returnOkTT[of "(=)"]
+lemmas corres_throwError_eq_same[corres_term] = corres_throwErrorTT[of "(=)"]
+
+lemma corres_get_trivial[corres_term]:
+  "corres_underlying sr nf nf' (\<lambda>s s'. (s,s') \<in> sr) \<top> \<top> get get"
+  by simp
+
+lemmas corres_underlying_stateAssert_stateAssert_trivial[corres_term] =
+  corres_underlying_stateAssert_stateAssert[where P=\<top> and P'=\<top>, simplified]
+
+lemmas corres_modify_tivial[corres_term] =
+  corres_modify[where P=\<top> and P'=\<top>, simplified]
```

Comment:
can you check on the var names this one and `corres_underlying_stateAssert_stateAssert_trivial` spit out? I'm having flashbacks to my monadic rewrite cleanup and some of the horrors `simplified` produced there.

Reply:
ah, yes, very good point. Probably better to write them out.